OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs

```diff
@@ -181,15 +158,81 @@ bool IsOnValidRampType()
 			return ramp == 0;
 		}
 
-		void GrantUpgrades()
+		/// <summary>Play deploy sound and animation.</summary>
+		void Deploy()
 		{
-			foreach (var up in info.Upgrades)
+			if (deployState != DeployState.Undeployed)
+				return;
+
+			if (!string.IsNullOrEmpty(info.DeploySound))
+				Game.Sound.Play(info.DeploySound, self.CenterPosition);
+
+			// Revoke upgrades that are used while undeployed.
+			RevokeUndeployedUpgrades();
+
+			// If there is no animation to play just grant the upgrades that are used while deployed.
+			// Alternatively, play the deploy animation and then grant the upgrades.
+			if (string.IsNullOrEmpty(info.DeployAnimation) || body.Value == null)
+				GrantDeployedUpgrades();
+			else
+				body.Value.PlayCustomAnimation(self, info.DeployAnimation, GrantDeployedUpgrades);
+		}
+
+		/// <summary>Play undeploy sound and animation and after that revoke the upgrades.</summary>
+		void Undeploy()
+		{
+			// Something went wrong, most likely due to deploy order spam and the fact that this is a delayed action.
```

Comment:
There's a comment here, but not on the check above.  Can we have a comment on both or on neither?